db/blob/blob_source.h

```diff
@@ -0,0 +1,139 @@
+#include "db/blob/blob_log_format.h"
+#include "rocksdb/rocksdb_namespace.h"
+#include "table/block_based/cachable_entry.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class Cache;
+struct ImmutableOptions;
+struct FileOptions;
+class HistogramImpl;
+class Status;
+class BlobFileReader;
+class FilePrefetchBuffer;
+class Slice;
+class IOTracer;
+
+class BlobSource {
+ public:
+  BlobSource(Cache* cache, const ImmutableOptions* immutable_options,
+             const FileOptions* file_options, const std::string& db_id,
+             const std::string& db_session_id, uint32_t column_family_id,
+             HistogramImpl* blob_file_read_hist,
+             const std::shared_ptr<IOTracer>& io_tracer);
+
+  BlobSource(const BlobSource&) = delete;
+  BlobSource& operator=(const BlobSource&) = delete;
+
+  ~BlobSource();
+
+  Status GetBlob(const ReadOptions& read_options, const Slice& user_key,
+                 uint64_t file_number, uint64_t offset, uint64_t file_size,
+                 uint64_t value_size, CompressionType compression_type,
+                 FilePrefetchBuffer* prefetch_buffer, PinnableSlice* value,
+                 uint64_t* bytes_read);
+
+  bool TEST_BlobInCache(uint64_t file_number, uint64_t file_size,
+                        uint64_t offset) const;
+
+ private:
+  CacheKey GetCacheKey(uint64_t file_number, uint64_t file_size,
+                       uint64_t offset) const;
+
+  void SaveValue(const Slice& src, PinnableSlice* dst);
+
+  Status MaybeReadBlobFromCache(FilePrefetchBuffer* prefetch_buffer,
+                                const ReadOptions& read_options,
+                                const CacheKey& cache_key, uint64_t offset,
+                                const bool wait,
+                                CachableEntry<Slice>* blob_entry) const;
+
+  Cache::Handle* GetEntryFromCache(const CacheTier& cache_tier,
+                                   Cache* blob_cache, const Slice& key,
+                                   const bool wait,
+                                   const Cache::CacheItemHelper* cache_helper,
+                                   const Cache::CreateCallback& create_cb,
+                                   Cache::Priority priority) const;
+
+  Status InsertEntryToCache(const CacheTier& cache_tier, Cache* blob_cache,
+                            const Slice& key,
+                            const Cache::CacheItemHelper* cache_helper,
+                            const Slice* blob, size_t charge,
+                            Cache::Handle** cache_handle,
+                            Cache::Priority priority) const;
+
+  Status GetBlobFromCache(const Slice& cache_key, Cache* blob_cache,
+                          CachableEntry<Slice>* blob, bool wait) const;
+
+  Status PutBlobToCache(const Slice& cache_key, Cache* blob_cache,
+                        CachableEntry<Slice>* cached_blob,
+                        PinnableSlice* blob) const;
+
+  static size_t SizeCallback(void* obj) {
```

Comment:
I suppose we don't need these either for now (since they are also cache tiering related); however, one thing to keep in mind down the line is that we would like to cache just the blob value, not the entire `BlobLogRecord`.

Reply:
Removed.